cpp/doc/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,739 @@
+
+(TODO: the above is not currently true for many modules.)
+
+Internal API headers containing `detail` namespace definitions that are used across translation
+units inside libcuspatial should be placed in `include/cuspatial/detail`.
+
+Note that (currently) header-only API files are in `include/cuspatial/experimental`, and their tests
+are in `tests/experimental`. When the header-only refactoring is complete these should be renamed or
+split into a separate library.
+
+## File extensions
+
+- `.hpp` : C++ header files
+- `.cpp` : C++ source files
+- `.cu`  : CUDA C++ source files
+- `.cuh` : Headers containing CUDA device code
+
+Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
+symbols that are only recognized by `nvcc`. Another indicator is Thrust algorithm APIs with a device
+execution policy (always `rmm::exec_policy` in libcuspatial).
+
+## Code and Documentation Style and Formatting
+
+libcuspatial code uses [snake_case](https://en.wikipedia.org/wiki/Snake_case) for all names except 
+in a few cases: template parameters, unit tests and test case names may use Pascal case, aka
+[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case). We do not use
+[Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), except sometimes when naming
+device data variables and their corresponding host copies (e.g. `d_data` and `h_data`). Private
+member variables are typically prefixed with an underscore. 
+
+Examples:
+
+```c++
+template <typename IteratorType>
+void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr)
+{
+  ...
+}
+
+class utility_class
+{
+  ...
+private:
+  int _rating{};
+  std::unique_ptr<cudf::column> _column{};
+}
+
+TYPED_TEST_SUITE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);
+
+TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
+{
+  ...
+}
+```
+
+C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your
+machine to use the `cuspatial/cpp/.clang-format` configuration file, and run `clang-format` on all
+changed code before committing it. The easiest way to do this is to configure your editor to
+"format on save", or to use `pre-commit`.
+
+Aspects of code style not discussed in this document and not automatically enforceable are typically
+caught during code review, or not enforced.
+
+### C++ Guidelines
+
+In general, we recommend following
+[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). We also
+recommend watching Sean Parent's [C++ Seasoning talk](https://www.youtube.com/watch?v=W2tWOdzgXHA),
+and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives."
+
+ * Prefer algorithms from STL and Thrust to raw loops.
+ * Prefer libcudf and RMM [owning data structures and views](#libcudf-data-structures) to raw
+   pointers and raw memory allocation.
```

Comment:
Corresponding the 3 rules, add one more addressing the synchronization bit.

```suggestion
 * Prefer algorithms from STL and Thrust to raw loops.
 * Prefer libcudf and RMM [owning data structures and views](#libcudf-data-structures) to raw
   pointers and raw memory allocation.
 * Prefer dispatching kernel to streams, instead of explicit synchronization.
```